### PR TITLE
Refactor Decimal128 averaging code to be vectorizable (and easier to read)

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/4973

# Rationale for this change
While working on a POC for new grouping code, I found I needed to call the decimal averaging code, but could not do so in a performant (vectorized) way because it created a `ScalarValue` 

# What changes are included in this PR?

1. Pull out the decimal average code into its own structure
2. Add comments that explain what is happening
3. Refactor `calculate_result_decimal_for_avg`

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/utils.rs [hunk 1]
```diff
@@ -37,45 +37,107 @@ pub fn get_accum_scalar_values_as_arrays(
         .collect::<Vec<_>>())
 }
 
-pub fn calculate_result_decimal_for_avg(
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/utils.rs:40]: This function is called once per output row in the current code. I want to be able to perform the setup calculations once and then call the minimal code per row in a loop. Thus this refactor

File: datafusion/physical-expr/src/aggregate/utils.rs [hunk 2]
```diff
@@ -37,45 +37,107 @@ pub fn get_accum_scalar_values_as_arrays(
...
+    }
+
+    /// Returns the `sum`/`count` as a i128 Decimal128 with
+    /// target_scale and target_precision and reporting overflow.
+    ///
+    /// * sum: The total sum value stored as Decimal128 with sum_scale
+    /// (passed to `Self::try_new`)
+    /// * count: total count, stored as a i128 (*NOT* a Decimal128 value)
+    #[inline(always)]
+    pub fn avg(&self, sum: i128, count: i128) -> Result<i128> {
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/utils.rs:99]: My goal is to call this function in a loop

**Review by viirya — COMMENTED** [review]
I'll review this soon or later in weekend.

**Comment by alamb** [comment]
> I'll review this soon or later in weekend.

Thanks @viirya  -- no rust from my perspective

**Review by Dandandan — APPROVED** [review]

File: datafusion/physical-expr/src/aggregate/utils.rs [hunk 3]
```diff
@@ -37,45 +37,107 @@ pub fn get_accum_scalar_values_as_arrays(
...
+    /// The maximum output value possible to represent with the target precision
+    target_max: i128,
+}
+
+impl Decimal128Averager {
+    /// Create a new `Decimal128Averager`:
+    ///
+    /// * sum_scale: the scale of `sum` values passed to [`Self::avg`]
+    /// * target_precision: the output precision
+    /// * target_precision: the output scale
```
Comment (viirya) [inline datafusion/physical-expr/src/aggregate/utils.rs:64]:
```suggestion
    /// * target_scale: the output scale
```

File: datafusion/physical-expr/src/aggregate/utils.rs (hunk 1 above)
Comment (viirya) [reply datafusion/physical-expr/src/aggregate/utils.rs:40]: Is it called per output row? Isn't it called in `evaluate`? I think `evaluate` is called only to get final aggregate value?

File: datafusion/physical-expr/src/aggregate/utils.rs (hunk 2 above)
Comment (viirya) [reply datafusion/physical-expr/src/aggregate/utils.rs:99]: I guess that you mean you will call this in a loop in new grouping code?

**Review by viirya — APPROVED** [review]
Refactoring looks good to me. I think the purpose is to call refactored function (`avg`) in other code (not this change) in vectorizable way?

**Comment by alamb** [comment]
> Refactoring looks good to me. I think the purpose is to call refactored function (avg) in other code (not this change) in vectorizable way?

That is right -- in case you are interested, the location is here (and it shows very promising results)

https://github.com/alamb/arrow-datafusion/blob/35b413239a1fdad1663231cddc73501748404835/datafusion/physical-expr/src/aggregate/average.rs#L172-L181

File: datafusion/physical-expr/src/aggregate/utils.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/utils.rs:40]: >  Isn't it called in evaluate? 

Yes

> I think evaluate is called only to get final aggregate value?

Correct -- it is called *for each group* to get the final aggregate value.  The ASCII art on this comment https://github.com/apache/arrow-datafusion/issues/4973#issuecomment-1608068287 might help to understand why

File: datafusion/physical-expr/src/aggregate/utils.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/utils.rs:99]: Exactly -- https://github.com/apache/arrow-datafusion/pull/6810#issuecomment-1616602087

**Pull request merged**